### ISSUE TITLE
Use new `device_lists_changes_in_room` table for fetching device lists in `/sync`

### ISSUE THREAD

**Issue by erikjohnston** [matrix-org/synapse#12388]
We added a table that tracks device list changes in a room in https://github.com/matrix-org/synapse/pull/12321. We can use this in `/sync` (and `/keys/changes`) to more efficiently calculate the device list changes without having to calculate the full set of users who share a room with the requester.

c.f. 

https://github.com/matrix-org/synapse/blob/573cd0f92f78de54cbe5dc9d0bf9f904f9d90011/synapse/handlers/sync.py#L1313-L1329

and

https://github.com/matrix-org/synapse/blob/573cd0f92f78de54cbe5dc9d0bf9f904f9d90011/synapse/handlers/device.py#L139-L143

The tricky thing is that since the table has been added recently we need to a) only use it once our minimum schema is `69` and b) only for device list stream IDs that have happened after a.

**Comment by chagai95** [comment]
Is this ready for testing? I could merge the PR on a fork and try it out. I have an account which can't log in to new devices...

**Comment by erikjohnston** [comment]
> Is this ready for testing? I could merge the PR on a fork and try it out. I have an account which can't log in to new devices...

Can you try running latest `develop`, https://github.com/matrix-org/synapse/pull/12321 and https://github.com/matrix-org/synapse/pull/12365 should hopefully fix logging in and I'd be interested in how they perform in real world use cases.

**Comment by chagai95** [comment]
I tried merging and I can see some merge conflicts I tried taking all the new stuff from the PR but some stuff was done later on develop so I'm not sure what to take:

```
<<<<<<< HEAD
                stream_ids,
                context,
            )

        async with self._device_list_id_gen.get_next_mult(
            len(device_ids)
        ) as stream_ids:
            await self.db_pool.runInteraction(
                "add_device_change_to_stream",
                add_device_changes_txn,
                stream_ids,
=======
                stream_ids_for_device_change,
                context,
                hosts_have_been_calculated=hosts is not None,
            )

            # If the set of hosts to send to has not been calculated yet (and so
            # `hosts` is None) or there are no `hosts` to send to, then skip
            # trying to persist them to the DB.
            if not hosts:
                return

            self._add_device_outbound_poke_to_stream_txn(
                txn,
                user_id,
                device_ids,
                hosts,
                stream_ids_for_outbound_pokes,
                context,
>>>>>>> b61c5c77cf72b47ca288660bcdc2a0d2cea7542f
```

**Comment by chagai95** [comment]
Sorry I'm having some trouble with building the docker image so I won't be able to test this soon...

**Comment by chagai95** [comment]
Is this a hidden config in the release and also in the tag? I'm not using the official releases, I have a fork and I merged...

**Comment by erikjohnston** [comment]
> Is this a hidden config in the release and also in the tag? I'm not using the official releases, I have a fork and I merged...

Oh sorry, I'm wrong, its not in 1.56. I'm getting the timeline wrong. It will be in 1.57 and then enabled by default in 1.58.

The commit you need is https://github.com/matrix-org/synapse/commit/5c9e39e6192e952ba8a5bb8e5485bc6067f91699, which is in the tag v1.57.0rc1.

**Comment by chagai95** [comment]
Hey @erikjohnston I managed to build a docker image and test but it seems like I still can't log in... Maybe I need to higher the cache factor? I'm using the defaults...

That's what @squahtx suggested here: https://github.com/matrix-org/synapse/issues/12429#issuecomment-1094903034

Here are the logs:
[synapse.log](https://github.com/matrix-org/synapse/files/8503876/synapse.log)

I found:

```
Apr 18 03:50:59 localhost.localdomain matrix-synapse[19576]: 2022-04-18 01:50:59,075 - synapse.http.server - 690 - WARNING - POST-12397 - Not sending response to request <XForwardedForRequest at 0x7fd634d36820 method='POST' uri='/_matrix/client/r0/login' clientproto='HTTP/1.0' site='8008'>, already disconnected.

```

in line 3425 after about an hour and so since clicking the login.

**Comment by richvdh** [comment]
I think this was fixed by #12365. @erikjohnston: if there is still outstanding work here, please can you clarify what it is.
